In [1]:
system("apt install git-lfs")
system("git clone https://github.com/zvictor12/R-statistics-spbu.git /content/R-statistics-spbu")

In [3]:
euroweight <- read.table("/content/R-statistics-spbu/lab2/datasets/euroweight.dat.txt", sep="\t", col.names = c("ID", "weight", "batch"))

In [9]:
install.packages("nortest")
library(nortest)

func1 <- function(x)
{
  lt <- lillie.test(x)
  print(lt)
  return(lt$p.val)
}

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [10]:
func1(euroweight$weight)


	Lilliefors (Kolmogorov-Smirnov) normality test

data:  x
D = 0.023354, p-value = 0.01311



[1] 0.01311357

In [11]:
for(i in 1:8)
{
  cat("BATCH", i)
  x <- euroweight[euroweight$batch == i, "weight"]
  func1(x)
}

BATCH 1
	Lilliefors (Kolmogorov-Smirnov) normality test

data:  x
D = 0.038472, p-value = 0.4901

BATCH 2
	Lilliefors (Kolmogorov-Smirnov) normality test

data:  x
D = 0.032659, p-value = 0.7434

BATCH 3
	Lilliefors (Kolmogorov-Smirnov) normality test

data:  x
D = 0.077923, p-value = 0.0008742

BATCH 4
	Lilliefors (Kolmogorov-Smirnov) normality test

data:  x
D = 0.045449, p-value = 0.2372

BATCH 5
	Lilliefors (Kolmogorov-Smirnov) normality test

data:  x
D = 0.034991, p-value = 0.6427

BATCH 6
	Lilliefors (Kolmogorov-Smirnov) normality test

data:  x
D = 0.0551, p-value = 0.06348

BATCH 7
	Lilliefors (Kolmogorov-Smirnov) normality test

data:  x
D = 0.042332, p-value = 0.3374

BATCH 8
	Lilliefors (Kolmogorov-Smirnov) normality test

data:  x
D = 0.070137, p-value = 0.004639



Как видим, вес монет в целом не подчиняется нормальному распределению, однако по отдельности во всех батчах кроме 3 и 8 вес монет нормален

In [13]:
B <- factor(euroweight$batch)
t <- pairwise.t.test(euroweight$weight, B)
print((t$p.value > 0.05)*1)

  1  2  3  4  5  6  7
2 1 NA NA NA NA NA NA
3 0  0 NA NA NA NA NA
4 0  1  0 NA NA NA NA
5 0  1  0  1 NA NA NA
6 1  1  1  0  0 NA NA
7 1  1  0  1  1  1 NA
8 1  1  1  0  0  1  1


Выше представлена матрица результатов сравнения мат. ожиданий с помощью t теста, видим что не у всех батчей мат. ожидания веса одинаковые

In [14]:
widths_vec <- c(8, 8, 8, 8)
babyboom <- read.fwf(
  file = "/content/R-statistics-spbu/lab1/datasets/babyboom.dat.txt",
  widths = widths_vec,
  col.names = c(
    "Time_of_birth",
    "Sex",
    "Birth_weight_in_grams",
    "Number_of_minutes_after_midnight_of_each_birth"
  ),
  strip.white = TRUE
)

In [19]:
t.test(babyboom[babyboom$Sex == 1, "Birth_weight_in_grams"], babyboom[babyboom$Sex == 2, "Birth_weight_in_grams"])


	Welch Two Sample t-test

data:  babyboom[babyboom$Sex == 1, "Birth_weight_in_grams"] and babyboom[babyboom$Sex == 2, "Birth_weight_in_grams"]
t = -1.4211, df = 27.631, p-value = 0.1665
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -593.1538  107.4273
sample estimates:
mean of x mean of y 
 3132.444  3375.308 


Принимаем нулевую гипотезу: мат. ож весов мальчиков и девочек равны

In [20]:
var.test(babyboom[babyboom$Sex == 1, "Birth_weight_in_grams"], babyboom[babyboom$Sex == 2, "Birth_weight_in_grams"])


	F test to compare two variances

data:  babyboom[babyboom$Sex == 1, "Birth_weight_in_grams"] and babyboom[babyboom$Sex == 2, "Birth_weight_in_grams"]
F = 2.1771, num df = 17, denom df = 25, p-value = 0.07526
alternative hypothesis: true ratio of variances is not equal to 1
95 percent confidence interval:
 0.9225552 5.5481739
sample estimates:
ratio of variances 
          2.177104 


Принимаем нулевую гипотезу: дисперсии весов мальчиков и девочек равны

In [21]:
iris <- read.table("/content/R-statistics-spbu/lab2/datasets/iris.txt", sep=",", col.names = c("Sepal_length", "Sepal_width", "Petal_length", "Petal_width", "Class"))

In [37]:
run_pairwise_test <- function(test) {
col_names <- c("Sepal_length", "Sepal_width", "Petal_length", "Petal_width")

for(i in col_names)
{
  class_names <- c("Iris-setosa", "Iris-versicolor", "Iris-virginica")
  result <- matrix(NA, nrow = 2, ncol = 2)
  rownames(result) <- class_names[-1]
  colnames(result) <- class_names[-3]
  cn <- class_names
  for(j in class_names[-3])
  {
    cn <- cn[cn != j]
    x <- iris[iris$Class == j, i]
    for(k in cn)
    {
      y <- iris[iris$Class == k, i]
      p <- test(x, y)
      result[k, j] <- ifelse(p$p.value > 0.05, 1, 0)
    }
  }
  cat("===== FEATURE", i, "=====\n\n")
  print(result)
  cat("\n\n")
}}

In [38]:
run_pairwise_test(ks.test)

===== FEATURE Sepal_length =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               0


===== FEATURE Sepal_width =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               0


===== FEATURE Petal_length =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               0


===== FEATURE Petal_width =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               0




Видим, что никакие классы не имеют одинаково распределенных попарно соответствующих признаков

In [39]:
run_pairwise_test(t.test)

===== FEATURE Sepal_length =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               0


===== FEATURE Sepal_width =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               0


===== FEATURE Petal_length =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               0


===== FEATURE Petal_width =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               0




In [40]:
run_pairwise_test(var.test)

===== FEATURE Sepal_length =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               1


===== FEATURE Sepal_width =====

                Iris-setosa Iris-versicolor
Iris-versicolor           1              NA
Iris-virginica            1               1


===== FEATURE Petal_length =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               1


===== FEATURE Petal_width =====

                Iris-setosa Iris-versicolor
Iris-versicolor           0              NA
Iris-virginica            0               0




Вывод показывает, что можно принять нулевую гипотезу о равенстве дисперсий всех классов у Sepal_width, а также у versicolor и virginica для petal и sepal length. Гипотеза о равенстве мат. ож. отклоняется везде

In [42]:
library(readxl)
file <- "/content/R-statistics-spbu/lab3/datasets/surgery.xlsx"
header_rows <- read_excel(file, range = cell_rows(1:2), col_names = FALSE)

fill_na <- function(x) {
  for(i in seq_along(x)) {
    if(is.na(x[i])) x[i] <- x[i-1]
  }
  return(x)
}

header_rows[1, ] <- fill_na(as.vector(header_rows[1, ]))
new_colnames <- paste(header_rows[1, ], header_rows[2, ], sep = "_")
new_colnames <- gsub(" ", "_", new_colnames)

surgery <- read_excel(file, skip = 2, col_names = FALSE)
colnames(surgery) <- new_colnames


New names:
• `` -> `...1`
• `` -> `...2`
• `` -> `...3`
• `` -> `...4`
New names:
• `` -> `...1`
• `` -> `...2`
• `` -> `...3`
• `` -> `...4`


In [62]:
surgery <- na.omit(surgery) #удалим строки с NA

In [63]:
success <- sum((surgery$After_operation_V_left > surgery$Before_operation_V_left) & (surgery$After_operation_V_right > surgery$Before_operation_V_right))

In [64]:
binom.test(success, nrow(surgery), p = 0.7)


	Exact binomial test

data:  success and nrow(surgery)
number of successes = 69, number of trials = 87, p-value = 0.06129
alternative hypothesis: true probability of success is not equal to 0.7
95 percent confidence interval:
 0.6928684 0.8725251
sample estimates:
probability of success 
             0.7931034 


In [65]:
binom.test(success, nrow(surgery), p = 0.8)


	Exact binomial test

data:  success and nrow(surgery)
number of successes = 69, number of trials = 87, p-value = 0.8933
alternative hypothesis: true probability of success is not equal to 0.8
95 percent confidence interval:
 0.6928684 0.8725251
sample estimates:
probability of success 
             0.7931034 


Обе гипотезы принимаются с двусторонней альтернативой, приблизительная оценка вероятности 0.793